### DISCUSSION TITLE
"Smart" way to apply default workspaces to multiple monitors (if multiple monitors exist)?

### DISCUSSION THREAD

**Discussion by jwhendy** [i3/i3#4784] [Q&A] [answered]
I recently upgraded my work from home setup with a dedicated external monitor. I read through [this section](https://i3wm.org/docs/userguide.html#workspace_screen) of the docs. The relevant example is:
```
workspace 1 output LVDS1
workspace 2 output primary
workspace 5 output VGA1 LVDS1
workspace "2: vim" output VGA1
```

Currently my laptop (DP-6) is on the left, external on the right (DP-0.1):
```
$ xrandr
Screen 0: minimum 8 x 8, current 3840 x 1080, maximum 32767 x 32767
DP-0.1 connected primary 1920x1080+1920+0 (normal left inverted right x axis y axis) 509mm x 286mm
   1920x1080     60.00*+
# [...]
DP-6 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 344mm x 193mm
   1920x1080     60.01*+  40.01 
```

By default, my external (primary) is getting ws_1 and laptop ws_2. I have a script where ctrl+shift+mod+arrow bumps the current container +/- one workspace, but this is having the effect of going: 1 (external) -> 2 (laptop) -> 3 (external)

What isn't clear from the settings is how I get something like this:

**when docked (and have the external monitor)**
- laptop stays on ws_1 and the external monitor can switch through ws_2 to ws_n
- this would let me put things on ws_1 for reference, but switch around easily with my shortcut between things I'm doing on the exernal monitor

**when not docked**
- laptop would get ws_1 and and the second bullet above would apply as well

Is this doable via some clever workspace definition per the docs example, or something I have to script by detecting my setup via e.g. xrandr and parsing the output?

Many thanks.

---

**Edit:** I did test the config suggested in the docs and this works as hard-coded. I assume this will just land me in ws_2 if I were to startx from laptop only. Again, not a huge deal, just wondered if someone with more knowledge would have a config idea that adapts to what's plugged in.
```
workspace $ws1 output DP-6
workspace $ws2 output DP-0.1
```

Request: could the docs be updated to clarify that a logout/startx cycle appears to be needed? I refreshed i3 after adding the above, restarted i3,  and saw no changes. After exiting and running startx again, the change was indeed applied.

**Comment by Airblader** [comment, marked as answer]
> Is this doable via some clever workspace definition

I'm not entirely sure I understood: is this about workspace placement (assign), or is it about your script moving containers to +/- one workspace? If it's the latter, I think your script just has to be a bit "smarter" to essentially achieve +/- one workspace on current output.

> Request: could the docs be updated to clarify that a logout/startx cycle appears to be needed? I refreshed i3 after adding the above, restarted i3, and saw no changes. After exiting and running startx again, the change was indeed applied.

That wouldn't be accurate. :-) `i3 reload` should suffice – however, this only applies to when a workspace is created, not retroactively to existing workspaces.

**Reply by jwhendy** [reply]
>...is this about workspace placement (assign), or is it about your script moving containers to +/- one workspace? 

This is about the way i3 by default handles the xrandr arrangement I showed (laptop left, external on right). The progression of ith workspace does not match the physical world (with location added for clarification)

> 1 (external/right-of-2) -> 2 (laptop/left-of-1) -> 3 (external/right-of-1)

Coupled with the fact that my external monitor may not always be plugged in, and therefore would have no need for $ws1 to be on the laptop and $ws2-n to be on the monitor. My case is relatively trivial, but this question could be of use to others with more complicated setups who would like a way to have i3 behave in one way when docked, and another when it's laptop screen only.

If the answer is "you can't, script it," understood. This discussion can still serve to communicate that more broadly.

> That wouldn't be accurate. :-) `i3 reload` should suffice...

I have `bindsym $mod+shift+r restart`, and was not familiar with `i3 reload`. I don't *think* the caveat applies, as I wasn't creating workspaces, just adding the two additional lines mentioned in my edit. I thought that a restart (which is how I apply other config changes) would swap them (because by default $ws1 was external, $ws2 was laptop).

In any case, the manual also doesn't call out `i3 reload` in this section on multiple displays. Looking back through the manual, maybe the reason I wasn't aware is that `reload` isn't called out at all in Section 2: Using i3, only restart is:

> 2.9. Restarting i3 inplace
> To restart i3 in place (and thus get into a clean state if there is a bug, or to upgrade to a newer version of i3) you can use $mod+Shift+r.

If you wanted to be more general vs. having redundant information on how users should reload (i.e. bad form to add "once you make changes, `reload`" in every section), perhaps [this section](https://i3wm.org/docs/userguide.html#_reloading_restarting_exiting) could be clarified?

>You can make i3 reload its configuration file with reload. You can also restart i3 inplace with the restart command to get it out of some weird state (if that should ever happen) or to perform an upgrade without having to restart your X session. 

After reading it, I thought "oh, I should have been using `reload` all this time." Yet, changing my config and issuing `$mod+shift+r` indeed reloads the config and picks up the changed bindsym I tweaked... so this description isn't entirely accurate. Anyway, all I'm really trying to say is "it wasn't clear to me what to do." 

There's precedent for calling out `reload` specifically now that I know what I'm looking for, though, so maybe a mention in 4.20 is reasonable?
> 5.3. Display mode
>The mode option can be changed during runtime through the bar mode command. On reload the mode will be reverted to its configured value.

Just thinking out loud. You built this and I get that it's hard to detach from the known answer. Sometimes there's value in hearing the perspective of a "fresh brain," as you can't unlearn what you already know like reflex :) 

Edits: "look through ~monitor~ manual", and clarified that by "it" I meant `reload`

**Reply by Airblader** [reply]
> If the answer is "you can't, script it," understood. 

I assume this is the answer, but I still haven't fully understood (admittedly my time thinking about this is limited, too). i3 does not name workspaces on its own, ever¹ – you do, when you create that workspace.

¹ There's a small lie in this – we name the first workspace per output because we need _some_ workspace, but we try to infer those names from your key bindings.

---

`reload` and `restart` are different things, but `restart` "includes" `reload`. If you feel that the docs can be improved, please feel free to send a PR and we can discuss the specific changes there.

My one comment here would be that throughout the entire userguide we cannot always refer to all variations of reload / restart, that would be counter-productive. Restarting is never "wrong"², often just unnecessary. But it's easier to tell a user to do too much than it is to explain nuances of this in unrelated sections.

² Of course there's an exception to this, too.

**Reply by jwhendy** [reply]
> I assume this is the answer, but I still haven't fully understood (admittedly my time thinking about this is limited, too). i3 does not name workspaces on its own, ever¹ – you do, when you create that workspace.

Ok. I likely followed some tutorial and have this in my config, if that's what you mean about me naming, not i3.

```
set $ws1 "1"
set $ws2 "2"
### [...]
set $ws10 "10"
```

Regardless of naming or not, it does seem that the manual says that their existence will match my experience?
> Also, the assignment of workspaces to screens will determine which workspace i3 uses for a new screen when adding screens or when starting (e.g., by default it will use 1 for the first screen, 2 for the second screen and so on).

My best attempt to simplify this is, how can I override this default behavior programmatically such that:
- if my external monitor exists, the order of the workspaces matches my physical world (with the first ws on the laptop, and subsequent ones on the external monitor)
- with no external monitor, the default behavior prevails

---

> `reload` and `restart` are different things, but `restart` "includes" `reload`.

Ok, in that case, my experience did not match. I can try to reproduce if you'd like. [My config](https://gist.github.com/jwhendy/fb2f752a0b64f52eb1d2546b13f40af2), including the section above defining workspace names, led to a default ws1 on my external monitor (right), and ws2 on my laptop (left). After adding the following to my config:
```
workspace $ws1 output DP-6     # laptop, formerly ws2
workspace $ws2 output DP-0.1   # external, formerly ws1
```
and issuing `restart`, nothing happened. Exiting and `startx` applied the expected change. Again, happy to reproduce this, perhaps in a week or so, if that would be helpful.

>But it's easier to tell a user to do too much than it is to explain nuances of this in unrelated sections.

Totally understood. Just that since `restart` didn't work for me and you said I should have done `reload` I actually assumed that `reload` was *more* comprehensive, so now I'm surprised to hear it's the other way around.

At present, the section doesn't say to do anything, and the more extreme thing I did failed to apply the change. Just taking the opportunity to provide feedback. I can create an issue if that's a better place for this, which could lead to a minor doc PR (which I'm also happy to propose).

**Reply by Airblader** [reply]
> I likely followed some tutorial and have this in my config, if that's what you mean about me naming, not i3. […] Regardless of naming or not, it does seem that the manual says that their existence will match my experience?

Yes, that is what I meant: those names are not defined by i3, but by you(r config). However, when i3 first starts up, it needs to show _some_ workspace (per screen), and those workspaces need _some_ name. So we do our best at guessing what you would want those names to be, by scanning the config file for certain configurations involving workspace names and picking from that list. Only if we can find nothing in there we'll fall back to coming up with names ourselves.

> After adding the following to my config: […] and issuing restart, nothing happened.

If the workspaces already existed, that is expected. Workspace assignments (even on `i3 restart`) do not evaluate on already existing workspaces, but only when that workspace is created (i.e., didn't exist before).

> Exiting and startx applied the expected change. 

That ends the i3 session entirely, so when you log back in every workspace is new to i3, and thus the assignments are evaluated.

> Just taking the opportunity to provide feedback. 

Appreciated!

> I can create an issue if that's a better place for this, which could lead to a minor doc PR (which I'm also happy to propose).

That'd be great; for a smaller docs PR you also don't need to create an issue necessarily, just submitting a PR would be fine!

**Reply by jwhendy** [reply]
We're aligned now on naming, but I think that's a slight tangent. Whether named via config or unnamed (and thus i3 comes up with names), is what I want only achievable via scripting in order to detect what is and is not currently connected? Or is there any way to put ws1 on the left given an external monitor, else on the laptop?

> If the workspaces already existed, that is expected.

Ah, ok, I went back to the top and I think I read this exactly opposite. 

> That wouldn't be accurate. :-) i3 reload should suffice – however, this only applies to when a workspace is created, not retroactively to existing workspaces.

I thought you were saying that `reload` *didn't* work for me since workspaces were being created, to which I replied I didn't think I was creating them. I now understand this to be exactly reverse: "if you change the layout for existing workspaces, there is no way to apply the changes other than stopping i3 and starting it again." Is that accurate?

I'll take a stab at some language that captures this discussion [edit: for the purpose of a doc PR], and I appreciate the assistance! I think all that's left it to confirm that the desired behavior is beyond what i3 can provide and thus requires a script, at which point I'll mark  as resolved. Thanks!

**Reply by Airblader** [reply]
> Or is there any way to put ws1 on the left given an external monitor, else on the laptop?

If you mean "if output X is connected, put it on whatever output is the most left in the RandR layout", no. If you mean "if output X is connected, put it on X, otherwise on Y", then it's just a precedence list:

```
workspace 1 output X Y
```

> I now understand this to be exactly reverse: "if you change the layout for existing workspaces, there is no way to apply the changes other than stopping i3 and starting it again." Is that accurate?

Another option would be to destroy that workspace (e.g. renaming it) and then recreating it, of course any windows on that workspace would now be on another one and would have to be moved.

I find the phrasing of that statement misleading for multiple reasons, though:

1. The "workspace layout" has a specific meaning in i3, and it is a different one; you're referring to workspace _assignments_.
2. The change _is_ applied when the config is reloaded, because the definition of a workspace assignment is that it affects newly created workspaces, not that it moves around existing workspaces.

**Reply by jwhendy** [reply]
> If you mean "if output X is connected, put it on whatever output is the most left in the RandR layout", no. If you mean "if output X is connected, put it on X, otherwise on Y", then it's just a precedence list:

Oh awesome! Yes, I think this will work great and bah, I saw the note about order... but somehow did not connect the dots about ordering them, with not connected just moving on to the next in the list. Thanks!

> The "workspace layout" has a specific meaning in i3, and it is a different one; you're referring to workspace assignments.

This also helps me understand better. Indeed, they existed and I was only looking to move the existing workspace.

Thanks for working through all of that. In re-skimming this thread, I think most of this was down to my description using terms in my mind vs. the canonical names/terms used by i3. Many thanks.
